File: rps/rps.js
/* eslint-disable max-lines-per-function */

const readline = require('readline-sync');
const CONFIG = require('./rock_paper_scissors.json');
const GAME_RULES = CONFIG.game_rules;
const VALID_MOVES = CONFIG.valid_moves;
const UTILITIES = require('./utilities');
const DISPLAY_ASSETS = require('./display_assets');
const GAMES_NEEDED_TO_WIN_MATCH = 2;


function createPlayer() {
  return {
    move: null,
    score: 0,
    moves: {},
    isGameWinner: false,

    getMove() {return this.move},
    setMove(value) {this.move = value},
    getScore() {return this.score},
    getMoves() {return this.moves},
    getIsGameWinner() {return this.isGameWinner},
    setIsGameWinner(value) {this.isGameWinner = value},
    isMatchWinner() {return this.score === GAMES_NEEDED_TO_WIN_MATCH},
    resetScore() {this.score = 0},

    winGame() {
      this.score += 1;
      this.isGameWinner = true;
    },

    updateMoves(move) {
      if (this.moves[move]) {
        this.moves[move] += 1;
      } else this.moves[move] = 1;
    },
  };
}

function createComputer() {
  let playerObject = createPlayer();

  let computerObject = {
    // calculateWeightedMoves determines which moves the computer should
    // favor based on the player's history
    calculateWeightedMoves(playerMoves, validMoves) {
      let weightedMoves = validMoves.slice();
      for (let move in playerMoves) {
        let moveCount = playerMoves[move];
        let moveDefeatedBy = GAME_RULES[move].defeated_by;
        for (let iterations = 0; iterations < moveCount; iterations += 1) {
          weightedMoves = weightedMoves.concat(moveDefeatedBy);
        }
      }
      return weightedMoves;
    },

    choose(playerMoves, validMoves = VALID_MOVES.rps) {
      let choices = this.calculateWeightedMoves(playerMoves, validMoves);
      let randomIndex = Math.floor(Math.random() * choices.length);
      this.move = choices[randomIndex];
    }
  };
  return Object.assign(playerObject, computerObject);
}

function createHuman() {
  let playerObject = createPlayer();

  let humanObject = {
    choose(_, validMoves = VALID_MOVES.rps) {
      let choice;
      console.clear();
      while (true) {
        console.log(`Please choose ${validMoves.join(', ')}:`);
        choice = readline.question();
        if (validMoves.includes(choice)) break;
        console.log('Sorry, invalid choice.');
      }
      this.move = choice;
    }
  };
  return Object.assign(playerObject, humanObject);
}

function createGameBoard(human, computer) {
  return {
    human: human,
    computer: computer,

    setHuman(value) {this.human = value},
    setComputer(value) {this.computer = value},

    combineAssets(...assets) {
      let assetsCombined = assets.map(elem => {
        return elem.split('\n');
      });

      let combinedDisplay = assetsCombined[0].map((elem, idx) => {
        let line = elem;
        for (let asset = 1; asset < assetsCombined.length; asset += 1) {
          line += assetsCombined[asset][idx];
        }
        return line + '\n';
      }).join('');

      return combinedDisplay;
    },

    displayMoves() {
      console.clear();
      let combinedDisplay = this.combineAssets(
        DISPLAY_ASSETS.yourMove + DISPLAY_ASSETS[this.human.getMove()],
        DISPLAY_ASSETS.computerMove + DISPLAY_ASSETS[this.computer.getMove()]
      );
      console.log(`${combinedDisplay}`);
    },

    displayWelcomeMessage() {
      console.clear();
      let combinedDisplay = this.combineAssets(
        DISPLAY_ASSETS.rock, DISPLAY_ASSETS.paper, DISPLAY_ASSETS.scissors
      );
      console.log(DISPLAY_ASSETS.welcomeRPS + combinedDisplay);
      console.log(`Win the match by being the first to win ${GAMES_NEEDED_TO_WIN_MATCH} games.`);
    },

    displayGoodbyeMessage() {
      console.log('Thanks for playing Rock, Paper, Scissors. Goodbye!');
    },

    displayGameTypes() {
      console.clear();
      let combinedDisplay = this.combineAssets(
        DISPLAY_ASSETS.lizard, DISPLAY_ASSETS.spock
      );
      console.log(DISPLAY_ASSETS.optionLS + combinedDisplay);
    },

    displayGameWinner() {
      if (this.human.getIsGameWinner()) {
        console.log(DISPLAY_ASSETS.youWin);
      } else if (this.computer.getIsGameWinner()) {
        console.log(DISPLAY_ASSETS.computerWins);
      } else {
        console.log(DISPLAY_ASSETS.tie);
      }
    }
  };
}

const RPSGame = {
  human: createHuman(),
  computer: createComputer(),
  gameBoard: createGameBoard(),
  validMoves: null,

  getValidMoves() {return this.validMoves},
  setValidMoves(value) {this.validMoves = value},

  resetGameBoard() {
    this.gameBoard.setHuman(this.human);
    this.gameBoard.setComputer(this.computer);
  },

  updatePlayerMoves() {
    this.human.updateMoves(this.human.move);
    this.computer.updateMoves(this.computer.move);
  },

  determineWinner() {
    let humanMove = this.human.move;
    let computerMove = this.computer.move;

    this.updatePlayerMoves();

    if (GAME_RULES[humanMove].defeats.includes(computerMove)) {
      this.human.winGame();
    } else if (GAME_RULES[computerMove].defeats.includes(humanMove)) {
      this.computer.winGame();
    }
  },

  playAgain() {
    UTILITIES.prompt('Next game? (y/n)');
    let answer = readline.question();
    return answer.toLowerCase()[0] === 'y';
  },

  newMatch() {
    UTILITIES.prompt('New match? (y/n)');
    let answer = readline.question();
    return answer.toLowerCase()[0] === 'y';
  },

  displayMatchSummary() {
    console.log(`Your win count: ${this.human.getScore()}`);
    console.log(`Computer win count: ${this.computer.getScore()}`);
  },

  isMatchOver() {
    return this.human.isMatchWinner() || this.computer.isMatchWinner();
  },

  displayMatchWinner() {
    if (this.human.isMatchWinner()) {
      console.log(DISPLAY_ASSETS.youWinMatch);
    } else if (this.computer.isMatchWinner()) {
      console.log(DISPLAY_ASSETS.computerWinsMatch);
    }
  },

  selectGameType() {
    this.gameBoard.displayGameTypes();
    UTILITIES.prompt('Would you like to include additional moves? (y/n)');
    let answer = readline.question();
    if (answer.toLowerCase()[0] === 'y') {
      this.setValidMoves(VALID_MOVES.rpsls);
    } else {
      this.setValidMoves(VALID_MOVES.rps);
    }
  },

  resetMatch() {
    this.human.resetScore();
    this.computer.resetScore();
    this.selectGameType();
  },

  resetGame() {
    this.human.setIsGameWinner(false);
    this.computer.setIsGameWinner(false);
  },

  continueGame() {
    readline.question('Click any key to continue:');
  },

  play() {
    this.gameBoard.displayWelcomeMessage();
    this.continueGame();
    while (true) {
      this.resetMatch();
      this.resetGameBoard();
      // Individual game loop
      while (true) {
        this.resetGame();
        this.human.choose(null, this.getValidMoves());
        this.computer.choose(this.human.getMoves(), this.getValidMoves());
        this.gameBoard.displayMoves();
        this.determineWinner();
        this.gameBoard.displayGameWinner();
        this.displayMatchSummary();
        if (this.isMatchOver()) break;
        if (!this.playAgain()) break;
      }
      this.displayMatchWinner();
      if (!this.newMatch()) break;
    }
    this.gameBoard.displayGoodbyeMessage();
  }
};

RPSGame.play();

File: rps/ruby/oo/rps.rb
module Affirmable
  def yes?(question, allowed_responses = ['y', 'n'])
    answer = ''
    loop do
      puts "#{question} (#{allowed_responses.join('/')})"
      answer = gets.chomp.downcase
      break unless !%w(y yes n no).include?(answer)
      puts "Please provide a valid response (#{allowed_responses.join('/')})"
    end
    answer.chars.first == 'y'
  end
end

class Move
  def self.valid_moves
    [Rock, Paper, Scissors, Lizard, Spock]
  end

  def self.display_valid_names
    valid_moves.map { |move| "#{move} (#{move.short_name})" }.join(', ')
  end

  def self.valid_names
    valid_moves.map { |move| [move.to_s.downcase, move.short_name] }.flatten
  end

  def self.move(name)
    valid_moves.each do |move|
      return move if move.to_s.downcase == name || move.short_name == name
    end
  end

  attr_reader :move
end

class Rock < Move
  def >(other_move)
    other_move.class == Scissors ||
      other_move.class == Lizard
  end

  def to_s
    'rock'
  end

  def self.short_name
    'r'
  end
end

class Paper < Move
  def >(other_move)
    other_move.class == Rock ||
      other_move.class == Spock
  end

  def to_s
    'paper'
  end

  def self.short_name
    'p'
  end
end

class Scissors < Move
  def >(other_move)
    other_move.class == Paper ||
      other_move.class == Lizard
  end

  def to_s
    'scissors'
  end

  def self.short_name
    'sc'
  end
end

class Lizard < Move
  def >(other_move)
    other_move.class == Spock ||
      other_move.class == Paper
  end

  def to_s
    'lizard'
  end

  def self.short_name
    'l'
  end
end

class Spock < Move
  def >(other_move)
    other_move.class == Scissors ||
      other_move.class == Rock
  end

  def to_s
    'spock'
  end

  def self.short_name
    'sp'
  end
end

class Player
  attr_accessor :move, :name
  attr_reader :score, :moves

  def initialize
    set_name
    @score = 0
    @moves = { Rock => 0, Paper => 0, Scissors => 0, Lizard => 0, Spock => 0 }
  end

  def record_win
    @score = score + 1
  end

  def reset_score
    @score = 0
  end

  def record_move
    @moves[@move.class] += 1
  end

  def display_moves
    moves.map { |move, count| "#{move.name}: #{count}" }.join(', ')
  end
end

class Human < Player
  def set_name
    system "clear"
    n = nil
    loop do
      puts "Type your name and hit return:"
      n = gets.chomp
      break unless n.empty?
      puts "Sorry, must enter a value."
    end
    self.name = n
  end

  def select_move
    system "clear"
    choice = nil
    loop do
      puts "Please choose #{Move.display_valid_names}:"
      choice = gets.downcase.chomp
      break if Move.valid_names.include? choice
      puts "Sorry, invalid choice."
    end
    Move.move(choice)
  end

  def choose
    self.move = select_move.new
  end
end

class Computer < Player
  def set_name
    @name = [R2D2, WallE, Skynet].sample
  end

  def choose
    choice = name::MOVE_PREFERENCE_WEIGHT.map { |k, v| [k] * v }.flatten.sample
    self.move = choice.new
  end
end

class R2D2 < Computer
  MOVE_PREFERENCE_WEIGHT = {
    Rock => 1,
    Paper => 0,
    Scissors => 0,
    Lizard => 0,
    Spock => 0
  }
end

class WallE < Computer
  MOVE_PREFERENCE_WEIGHT = {
    Rock => 0,
    Paper => 1,
    Scissors => 1,
    Lizard => 1,
    Spock => 1
  }
end

class Skynet < Computer
  MOVE_PREFERENCE_WEIGHT = {
    Rock => 0,
    Paper => 0,
    Scissors => 1,
    Lizard => 0,
    Spock => 2
  }
end

class RPSGame
  attr_reader :human, :computer

  def initialize(human, computer)
    @human = human
    @computer = computer
  end

  def display_moves
    system "clear"
    puts "#{human.name} chose #{human.move}."
    puts "#{computer.name} chose #{computer.move}."
    puts ""
  end

  def display_game_winner
    if human.move > computer.move
      puts "#{human.name} won!"
    elsif computer.move > human.move
      puts "#{computer.name} won!"
    else
      puts "It's a tie!"
    end
  end

  def play
    human.choose
    human.record_move
    computer.choose
    computer.record_move
    display_moves
    display_game_winner
  end
end

class RPSMatch
  GAMES_TO_WIN_MATCH = 2
  include Affirmable

  attr_reader :game, :human, :computer

  def initialize
    @human = Human.new
    @computer = Computer.new
    @game = RPSGame.new(human, computer)
  end

  def display_welcome_message
    system "clear"
    puts "*****************************************************"
    puts "Welcome to Rock, Paper, Scissors, Lizard, Spock!"
    puts "*****************************************************"
    puts "You must win #{GAMES_TO_WIN_MATCH} games to win the match!"
    puts ""
  end

  def display_goodbye_message
    puts "Thanks for playing. Goodbye!"
  end

  def display_match_score
    human_score = human.score
    computer_score = computer.score
    puts "#{human.name} has won " \
     "#{human_score} game#{human_score == 1 ? '' : 's'}"
    puts "#{computer.name} has won " \
     "#{computer_score} game#{computer_score == 1 ? '' : 's'}"
    puts ""
  end

  def match_winner?
    human.score == GAMES_TO_WIN_MATCH ||
      computer.score == GAMES_TO_WIN_MATCH
  end

  def display_match_winner
    if human.score == GAMES_TO_WIN_MATCH
      puts "#{human.name} won the match!"
    elsif computer.score == GAMES_TO_WIN_MATCH
      puts "#{computer.name} won the match!"
    end
  end

  def display_move_history
    system "clear"
    puts "#{human.name}'s move history:\n #{human.display_moves}"
    puts "#{computer.name}'s move history:\n #{computer.display_moves}"
    puts ""
  end

  def play_another_game?
    yes?("Would you like to play another game?")
  end

  def play_another_match?
    yes?("Would you like to play another match?")
  end

  def see_move_history?
    system "clear"
    yes?("Would you like to see the move history?")
  end

  def continue?
    puts "Tap return when you're ready to play..."
    gets
  end

  def record_game_winner
    if human.move > computer.move
      human.record_win
    elsif computer.move > human.move
      computer.record_win
    end
  end

  def reset_game
    human.reset_score
    computer.reset_score
  end

  def game_loop
    loop do
      game.play
      record_game_winner
      display_match_score
      break if match_winner? || !play_another_game?
    end
  end

  def match_loop
    loop do
      game_loop
      break unless match_winner? # exited before match ended, terminate game
      display_match_winner
      break unless play_another_match?
      reset_game
    end
  end

  def play
    display_welcome_message
    continue?
    match_loop
    display_move_history if see_move_history?
    display_goodbye_message
  end
end

RPSMatch.new.play
